keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredSecurityGroupVeto.kt

```diff
@@ -0,0 +1,115 @@
+import com.netflix.spinnaker.keel.api.ec2.LoadBalancerSpec
+import com.netflix.spinnaker.keel.api.ec2.OverrideableClusterDependencyContainer
+import com.netflix.spinnaker.keel.api.ec2.securityGroupsByRegion
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverCache
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.core.api.DEFAULT_SERVICE_ACCOUNT
+import com.netflix.spinnaker.keel.ec2.CLOUD_PROVIDER
+import com.netflix.spinnaker.keel.veto.Veto
+import com.netflix.spinnaker.keel.veto.VetoResponse
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import retrofit2.HttpException
+
+@Component
+class RequiredSecurityGroupVeto(
+  private val cloudDriver: CloudDriverService,
+  private val cloudDriverCache: CloudDriverCache
+) : Veto {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override suspend fun check(resource: Resource<*>) =
+    resource.spec.toSecurityGroupDependencies()
+      ?.let {
+        val missingSecurityGroupRegions = checkSecurityGroups(it)
+        VetoResponse(
+          allowed = missingSecurityGroupRegions.isEmpty(),
+          vetoName = name(),
+          message = missingSecurityGroupRegions.entries.joinToString(separator = "\n") { (securityGroup, missingRegions) ->
+            "Security group $securityGroup is not found in ${missingRegions.joinToString()}"
+          }
+        )
+      } ?: VetoResponse(true, name())
+
+  private suspend fun checkSecurityGroups(spec: SecurityGroupDependencies): Map<String, List<String>> {
+    val jobs = mutableListOf<Job>()
+    val securityGroupMissingRegions = mutableMapOf<String, MutableList<String>>()
+    supervisorScope {
+      spec.securityGroupsInRegions.forEach { (securityGroupName, regions) ->
+        regions.forEach { region ->
+          launch {
+            if (!securityGroupExists(spec.account, spec.subnet, region, securityGroupName)) {
+              with(securityGroupMissingRegions) {
+                putIfAbsent(securityGroupName, mutableListOf())
+                getValue(securityGroupName).add(region)
+              }
+            }
+          }
+            .also { jobs.add(it) }
+        }
+      }
+    }
+
+    jobs.forEach { it.join() }
+
+    return securityGroupMissingRegions
+  }
+
+  private suspend fun securityGroupExists(
+    account: String,
+    vpcName: String?,
+    region: String,
+    name: String
+  ): Boolean =
+    runCatching {
+      cloudDriver.getSecurityGroup(
+        user = DEFAULT_SERVICE_ACCOUNT,
+        account = account,
+        type = CLOUD_PROVIDER,
+        securityGroupName = name,
+        region = region,
+        vpcId = if (vpcName == null) null else cloudDriverCache.networkBy(vpcName, account, region).id
+      )
+    }
+      .onFailure { ex ->
+        if (ex is HttpException && ex.code() == 404) {
```

Comment:
Is this a subclass of `RetrofitError` when the error is at the HTTP layer? I hadn't seen this before, but I ask because I could simplify code I wrote elsewhere to do similar things...

Reply:
Yeah, it is. I think it's any `RetrofitError` where you got an HTTP status code (i.e. not a socket closing or that kind of thing).